### ISSUE TITLE
NoMethodError: undefined method `id' for nil:NilClass "devise_masquerade_#{masqueraded_resource_name}_#{masqueradable_resource.id}_#{guid}".to_sym

### ISSUE THREAD

**Issue by rodriguevara** [oivoodoo/devise_masquerade#140]
Hi guys, thanks for mantaining this gem! I'm getting this error:

```
NoMethodError: undefined method `id' for nil:NilClass
    "devise_masquerade_#{masqueraded_resource_name}_#{masqueradable_resource.id}_#{guid}".to_sym
```
 sometimes when trying to log in. I'm using the `masquerade_path` method. It happens sometimes, and I couldn't find a 'pattern' so to speak. Most of the times works, but with some users this error comes up. Any ideas? Thanks a lot!

**Comment by adamjleonard** [comment]
Just started happening out of the blue for me as well

**Comment by rodriguevara** [comment]
@adamjleonard did you find a solution?

**Comment by adamjleonard** [comment]
@rodriguevara I did not

**Comment by oivoodoo** [comment]
Hi @adamjleonard and @rodriguevara 

have you tried to switch cache to session storage method?

```
    Devise.masquerade_storage_method = :session
```

**Comment by rodriguevara** [comment]
Hi @oivoodoo , thanks for replying. Where should this code go?

**Comment by oivoodoo** [comment]
I don't remember Rails enough :0 but I believe in place where you have devise initializer.

**Comment by rodriguevara** [comment]
@oivoodoo That didnt help, but I noticed that the error occurs after a minute of staying in the view where I call the masquerade, let's say my users index. If a minute pass without me doing anything else, then when trying to masquerade, the error comes up. So maybe it has something to do with `@@masquerade_expires_in = 1.minute`?

**Comment by oivoodoo** [comment]
@rodriguevara it's the normal behaviour, you can increase the time but lets say if your admin went for coffee and keep the page opened with masquerade access, we need to make possible to skip ability to anyone to do the authentication.

**Comment by rodriguevara** [comment]
@oivoodoo yep, that makes sense, it would be helpful a rescue for that error, maybe redirecting to root_path or something like that

**Comment by umaimaz123** [comment]
@rodriguevara  did this help you or you find the right solution?

**Comment by rodriguevara** [comment]
@umaimaz123 `@@masquerade_expires_in = 1.minute` you need to increase the time. 10 minutes works just fine for me

**Comment by remy727** [comment]
My solution:

- Step1: Add `id` query param on view like `masquerade_path(user, id: user.id)`.
- Step 2: Override `find_masqueradable_resource` method.
```
def find_masqueradable_resource
  GlobalID::Locator.locate_signed(params[Devise.masquerade_param], for: 'masquerade') ||
    masqueraded_resource_class.find(params[:id])
end
```

**Comment by oivoodoo** [comment]
Hi @remy727 .

Could you add some notes why it would help? would it be possible to include as feature scenario / spec to the library your changes as enhancements? probably it would help the other developers as well. 

thank you!

**Comment by fabiosimoesds** [comment]
This also started to happen to me.
It seems like a problem with the routes

These are my routes before the upgrade:
![CleanShot 2024-10-15 at 12 23 27@2x](https://github.com/user-attachments/assets/f9f281bf-5d65-4ddc-a79b-382fdf877e1c)

These are my routes after the upgrade:
![CleanShot 2024-10-15 at 12 24 25@2x](https://github.com/user-attachments/assets/da154b07-5658-4f53-b7c4-a18f340cd087)

No /:id in params.

**Comment by Kulgar** [comment]
@oivoodoo : I used @remy727 solution today. 

The thing is that the code `GlobalID::Locator.locate_signed(params[Devise.masquerade_param], for: 'masquerade')` returns nil after reaching the `@@masquerade_expires_in` configuration. 

So, passing the ID of the masqueraded ressource in the params and using that to retrieve the ressource in find_masqueradable_resource if the above code returns nil fix the problem. 

In fact that is quite the same logic you did in [your documentation for FriendlyId](https://github.com/oivoodoo/devise_masquerade?tab=readme-ov-file#overriding-the-finder). If we follow your documentation when using friendly ID, we won't have the above problem either, as that problem is related to the fact that the locate_signed method returns nil. 

So, I wonder why this method returns nil after "masquerade_expires_in" has been reached ? 

`GlobalID::Locator.locate_signed(params[Devise.masquerade_param], for: 'masquerade')` 

And is that really expected?

In any case the solution of @remy727 works. Coupled with Pundit/Cancancan, and devise timeout, I think it's ok at the security level to use that solution, don't you think?
